src/core/api.rs

```diff
@@ -336,6 +356,188 @@ impl EncryptedHeader {
 /// - `metadata`        : additional data symmetrically encrypted in a header
 #[derive(Debug, PartialEq, Eq)]
 pub struct CleartextHeader {
-    pub symmetric_key: SymmetricKey<SEED_LENGTH>,
+    pub seed: Secret<SEED_LENGTH>,
     pub metadata: Option<Vec<u8>>,
 }
+
+pub trait CovercryptKEM {
+    /// Generates an encapsulation for the given access
+    /// policy.
+    ///
+    /// # Error
+    ///
+    /// Returns an error if the access policy is not valid.
+    fn encaps(
+        &self,
+        mpk: &MasterPublicKey,
+        policy: &Policy,
+        ap: &str,
+    ) -> Result<(Secret<SEED_LENGTH>, Encapsulation), Error>; // todo : symmetric_key or seed?
+
+    /// Attempts opening the given encapsulation using the given
+    /// user secret key.
+    ///
+    /// Returns the encapsulated symmetric key if the user key holds
+    /// the correct rights.
+    fn decaps(
+        &self,
+        usk: &UserSecretKey,
+        enc: Encapsulation,
+    ) -> Result<Option<Secret<SEED_LENGTH>>, Error>;
+}
+
+impl CovercryptKEM for Covercrypt {
+    /// Generates an encapsulation for the given access
+    /// policy.
+    ///
+    /// # Error
+    ///
+    /// Returns an error if the access policy is not valid.
+    fn encaps(
+        &self,
+        mpk: &MasterPublicKey,
+        policy: &Policy,
+        ap: &str,
+    ) -> Result<(Secret<SEED_LENGTH>, Encapsulation), Error> {
+        let ap = &AccessPolicy::parse(ap)?;
+        encaps(
+            &mut *self.rng.lock().expect("Mutex lock failed!"),
+            mpk,
+            &policy.generate_point_coordinates(ap.clone())?,
+        )
```

Comment:
Another example of using a variable to store a reference tricking you into a useless clone.
```suggestion
        let ap = AccessPolicy::parse(ap)?;
        encaps(
            &mut *self.rng.lock().expect("Mutex lock failed!"),
            mpk,
            &policy.generate_point_coordinates(ap)?,
        )
```